src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleSerializer.cs

```diff
@@ -0,0 +1,151 @@
+        public BundleSerializer()
+        {
+        }
+
+        public async Task Serialize(Hl7.Fhir.Model.Bundle bundle, Stream outputStream)
+        {
+            using (RecyclableMemoryStream ms = new RecyclableMemoryStream(_memoryStreamManager))
+            using (Utf8JsonWriter writer = new Utf8JsonWriter(ms, _writerOptions))
+            using (StreamWriter streamWriter = new StreamWriter(ms))
+            {
+                writer.WriteStartObject();
+
+                writer.WriteString("resourceType", bundle.ResourceType.GetLiteral());
+                writer.WriteString("id", bundle.Id);
+
+                SerializeMetadata();
+
+                writer.WriteString("type", bundle.Type?.GetLiteral());
+
+                SerializeLinks();
+
+                if (bundle.Total.HasValue)
+                {
+                    writer.WriteNumber("total", bundle.Total.Value);
+                }
+
+                await SerializeEntries();
+
+                writer.WriteEndObject();
+                await writer.FlushAsync();
+
+                ms.WriteTo(outputStream);
+
+                void SerializeMetadata()
+                {
+                    if (bundle.Meta != null)
+                    {
+                        writer.WriteStartObject("meta");
+                        writer.WriteString("lastUpdated", bundle.Meta?.LastUpdated?.ToString(DateTimeOffsetFormat));
+                        writer.WriteEndObject();
+                    }
+                }
+
+                void SerializeLinks()
+                {
+                    if (bundle.Link?.Any() == true)
+                    {
+                        writer.WriteStartArray("link");
+
+                        foreach (var link in bundle.Link)
+                        {
+                            writer.WriteStartObject();
+                            writer.WritePropertyName("relation");
+                            writer.WriteStringValue(link.Relation);
+                            writer.WritePropertyName("url");
+                            writer.WriteStringValue(link.Url);
+                            writer.WriteEndObject();
+                        }
+
+                        writer.WriteEndArray();
+                    }
+                }
+
+                async Task SerializeEntries()
+                {
+                    if (bundle.Entry?.Any() == true)
+                    {
+                        writer.WriteStartArray("entry");
+                        foreach (var entry in bundle.Entry)
+                        {
+                            var rawBundleEntry = entry as RawBundleEntryComponent;
+                            writer.WriteStartObject();
+
+                            if (!string.IsNullOrEmpty(rawBundleEntry.FullUrl))
+                            {
+                                writer.WriteString("fullUrl", rawBundleEntry.FullUrl);
+                            }
+
+                            writer.WritePropertyName("resource");
```

Comment:
Can we avoid buffering by instead manually writing something like ", resource = " instead?

Reply:
Should be resolved now. It is nicer than having to copy over to another stream, but it is more error prone and makes me feel a bit uneasy - see comment about the comma :)